File: BDFM-RAG-System/src/services/sync.service.ts
import databaseService from './database.service';
import embeddingService from './embedding.service';
import qdrantService from './qdrant.service';
import logger from '../utils/logger';
import config from '../config';
import { SyncRequest, SyncResult, Correspondence } from '../models';
import { formatDuration } from '../utils/helpers';

export class SyncService {
  /**
   * Synchronize correspondences from PostgreSQL to Qdrant
   */
  async syncCorrespondences(request: SyncRequest): Promise<SyncResult> {
    const startTime = Date.now();
    let synced = 0;
    let failed = 0;
    const errors: string[] = [];

    try {
      logger.info(
        `Starting ${request.type} sync of correspondences with batch size ${request.batchSize || 100}`
      );

      // Get correspondences based on sync type
      let correspondences: Correspondence[];

      if (request.type === 'incremental' && request.fromDate) {
        const sinceDate = new Date(request.fromDate);
        correspondences = await databaseService.getCorrespondencesSince(
          sinceDate
        );
        logger.info(
          `Found ${correspondences.length} correspondences updated since ${request.fromDate}`
        );
      } else {
        // Full sync - get all correspondences in batches
        const batchSize = request.batchSize || 100;
        correspondences = [];
        let offset = 0;
        let hasMore = true;

        while (hasMore) {
          const batch = await databaseService.getAllCorrespondences(
            batchSize,
            offset
          );

          if (batch.length === 0) {
            hasMore = false;
          } else {
            correspondences.push(...batch);
            offset += batchSize;
            logger.info(
              `Fetched batch: ${batch.length} correspondences (total: ${correspondences.length})`
            );
          }
        }
      }

      logger.info(
        `Processing ${correspondences.length} correspondences for embedding`
      );

      // Process in batches
      const batchSize = request.batchSize || 100;
      const totalBatches = Math.ceil(correspondences.length / batchSize);

      for (let i = 0; i < correspondences.length; i += batchSize) {
        const batch = correspondences.slice(i, i + batchSize);
        const currentBatch = Math.floor(i / batchSize) + 1;

        logger.info(
          `Processing batch ${currentBatch}/${totalBatches} (${batch.length} items)`
        );

        try {
          await this.processBatch(batch);
          synced += batch.length;
        } catch (error: any) {
          logger.error(`Error processing batch ${currentBatch}:`, error);
          failed += batch.length;
          errors.push(
            `Batch ${currentBatch}: ${error.message || 'Unknown error'}`
          );
        }
      }

      const duration = Date.now() - startTime;

      logger.info(
        `Sync completed: ${synced} synced, ${failed} failed, duration: ${formatDuration(duration)}`
      );

      return {
        success: failed === 0,
        synced,
        failed,
        duration: formatDuration(duration),
        errors: errors.length > 0 ? errors : undefined,
      };
    } catch (error: any) {
      const duration = Date.now() - startTime;
      logger.error('Fatal error during sync:', error);

      return {
        success: false,
        synced,
        failed: failed + 1,
        duration: formatDuration(duration),
        errors: [error.message || 'Unknown fatal error'],
      };
    }
  }

  /**
   * Process a batch of correspondences
   */
  private async processBatch(correspondences: Correspondence[]): Promise<void> {
    for (const correspondence of correspondences) {
      try {
        // Delete old embeddings for this correspondence first to prevent duplicates
        await qdrantService.deleteByCorrespondenceId(
          config.collections.correspondence,
          correspondence.id
        );

        // Generate embeddings
        const embeddings =
          await embeddingService.generateEmbeddingsForCorrespondence(
            correspondence
          );

        // Upsert to Qdrant (now with deterministic IDs)
        await qdrantService.upsertEmbeddings(
          config.collections.correspondence,
          embeddings
        );

        logger.debug(
          `Synced correspondence ${correspondence.id} (${correspondence.mailNum})`
        );
      } catch (error: any) {
        logger.error(
          `Error syncing correspondence ${correspondence.id}:`,
          error
        );
        throw error; // Re-throw to be caught by batch handler
      }
    }
  }

  /**
   * Sync a single correspondence
   */
  async syncSingleCorrespondence(correspondenceId: string): Promise<boolean> {
    try {
      logger.info(`Syncing single correspondence: ${correspondenceId}`);

      const correspondence =
        await databaseService.getCorrespondenceById(correspondenceId);

      if (!correspondence) {
        logger.warn(`Correspondence ${correspondenceId} not found`);
        return false;
      }

      // Delete old embeddings for this correspondence first to prevent duplicates
      await qdrantService.deleteByCorrespondenceId(
        config.collections.correspondence,
        correspondenceId
      );

      // Generate embeddings
      const embeddings =
        await embeddingService.generateEmbeddingsForCorrespondence(
          correspondence
        );

      // Upsert to Qdrant (now with deterministic IDs)
      await qdrantService.upsertEmbeddings(
        config.collections.correspondence,
        embeddings
      );

      logger.info(`Successfully synced correspondence ${correspondenceId}`);
      return true;
    } catch (error) {
      logger.error(`Error syncing correspondence ${correspondenceId}:`, error);
      return false;
    }
  }

  /**
   * Delete correspondence from Qdrant
   */
  async deleteCorrespondence(correspondenceId: string): Promise<boolean> {
    try {
      logger.info(`Deleting correspondence ${correspondenceId} from Qdrant`);

      await qdrantService.deleteByCorrespondenceId(
        config.collections.correspondence,
        correspondenceId
      );

      logger.info(`Successfully deleted correspondence ${correspondenceId}`);
      return true;
    } catch (error) {
      logger.error(
        `Error deleting correspondence ${correspondenceId}:`,
        error
      );
      return false;
    }
  }

  /**
   * Get sync statistics
   */
  async getSyncStatistics(): Promise<any> {
    try {
      const postgresCount = await databaseService.getCorrespondenceCount();
      const qdrantCount = await qdrantService.getCollectionCount(
        config.collections.correspondence
      );

      return {
        postgresql: {
          total: postgresCount,
        },
        qdrant: {
          total: qdrantCount,
        },
        synced: Math.min(postgresCount, qdrantCount),
        pending: Math.max(0, postgresCount - qdrantCount),
      };
    } catch (error) {
      logger.error('Error getting sync statistics:', error);
      throw error;
    }
  }

  /**
   * Rebuild index (delete and re-sync all)
   */
  async rebuildIndex(): Promise<SyncResult> {
    try {
      logger.info('Rebuilding index: deleting collection...');

      // Delete and recreate collection
      await qdrantService.deleteCollection(config.collections.correspondence);
      await qdrantService.initialize();

      logger.info('Collection recreated, starting full sync...');

      // Perform full sync
      return await this.syncCorrespondences({
        type: 'full',
        batchSize: 100,
      });
    } catch (error: any) {
      logger.error('Error rebuilding index:', error);
      throw error;
    }
  }

  /**
   * Remove duplicate embeddings for all correspondences
   * This function identifies and removes old duplicates that were created with random UUIDs
   */
  async removeDuplicates(): Promise<{
    success: boolean;
    processed: number;
    cleaned: number;
    message: string;
  }> {
    try {
      logger.info('Starting duplicate removal process...');

      // Get all correspondences from PostgreSQL
      const correspondences: string[] = [];
      let offset = 0;
      const batchSize = 100;
      let hasMore = true;

      while (hasMore) {
        const batch = await databaseService.getAllCorrespondences(
          batchSize,
          offset
        );

        if (batch.length === 0) {
          hasMore = false;
        } else {
          correspondences.push(...batch.map((c) => c.id));
          offset += batchSize;
        }
      }

      logger.info(
        `Found ${correspondences.length} correspondences to check for duplicates`
      );

      let cleaned = 0;

      // For each correspondence, delete all embeddings and re-sync
      // This will remove old UUID-based embeddings and create new deterministic ones
      for (const correspondenceId of correspondences) {
        try {
          // Delete all embeddings for this correspondence
          await qdrantService.deleteByCorrespondenceId(
            config.collections.correspondence,
            correspondenceId
          );

          // Re-sync with deterministic IDs
          await this.syncSingleCorrespondence(correspondenceId);
          cleaned++;

          if (cleaned % 50 === 0) {
            logger.info(`Cleaned ${cleaned}/${correspondences.length} correspondences`);
          }
        } catch (error: any) {
          logger.warn(
            `Failed to clean correspondence ${correspondenceId}: ${error.message}`
          );
          // Continue with next correspondence
        }
      }

      logger.info(
        `Duplicate removal completed: ${cleaned}/${correspondences.length} correspondences processed`
      );

      return {
        success: true,
        processed: correspondences.length,
        cleaned,
        message: `Successfully removed duplicates for ${cleaned} correspondences`,
      };
    } catch (error: any) {
      logger.error('Error removing duplicates:', error);
      return {
        success: false,
        processed: 0,
        cleaned: 0,
        message: `Failed to remove duplicates: ${error.message}`,
      };
    }
  }
}

// Export singleton instance
export default new SyncService();

File: BDFM-RAG-System/bdfm-rag-python/app/services/sync_service.py
"""Sync service for syncing correspondences from PostgreSQL to Qdrant"""
from typing import Dict, Any, List
import time
from datetime import datetime
from app.config import settings
from app.utils.logger import logger
from app.utils.helpers import format_duration
from app.models import SyncRequest, SyncResult
from app.database.postgres import database_service
from app.services.embedding_service import embedding_service
from app.services.qdrant_service import qdrant_service


class SyncService:
    """Service for syncing correspondences to Qdrant"""
    
    async def sync_correspondences(self, request: SyncRequest) -> SyncResult:
        """
        مزامنة المراسلات من PostgreSQL إلى Qdrant
        
        Args:
            request: sync request
            
        Returns:
            نتيجة المزامنة
        """
        start_time = time.time()
        synced = 0
        failed = 0
        errors = []
        
        try:
            logger.info(f"Starting {request.type} sync with batch size {request.batch_size}")
            
            # Get correspondences based on sync type
            if request.type == "incremental" and request.from_date:
                since_date = datetime.fromisoformat(request.from_date)
                correspondences = await database_service.get_correspondences_since(since_date)
                logger.info(f"Found {len(correspondences)} correspondences updated since {request.from_date}")
            else:
                # Full sync - get all correspondences in batches
                batch_size = request.batch_size or 100
                correspondences = []
                offset = 0
                has_more = True
                
                while has_more:
                    batch = await database_service.get_all_correspondences(batch_size, offset)
                    if not batch:
                        has_more = False
                    else:
                        correspondences.extend(batch)
                        offset += batch_size
                        logger.info(f"Fetched batch: {len(batch)} correspondences (total: {len(correspondences)})")
            
            logger.info(f"Processing {len(correspondences)} correspondences for embedding")
            
            # Process in batches
            batch_size = request.batch_size or 100
            total_batches = (len(correspondences) + batch_size - 1) // batch_size
            
            for i in range(0, len(correspondences), batch_size):
                batch = correspondences[i:i + batch_size]
                current_batch = i // batch_size + 1
                
                logger.info(f"Processing batch {current_batch}/{total_batches} ({len(batch)} items)")
                
                try:
                    await self._process_batch(batch)
                    synced += len(batch)
                except Exception as e:
                    logger.error(f"Error processing batch {current_batch}: {e}")
                    failed += len(batch)
                    errors.append(f"Batch {current_batch}: {str(e)}")
            
            duration = time.time() - start_time
            
            logger.info(f"Sync completed: {synced} synced, {failed} failed, duration: {format_duration(duration * 1000)}")
            
            return SyncResult(
                success=failed == 0,
                synced=synced,
                failed=failed,
                duration=format_duration(duration * 1000),
                errors=errors if errors else None
            )
            
        except Exception as e:
            duration = time.time() - start_time
            logger.error(f"Fatal error during sync: {e}")
            
            return SyncResult(
                success=False,
                synced=synced,
                failed=failed + 1,
                duration=format_duration(duration * 1000),
                errors=[str(e)]
            )
    
    async def _process_batch(self, correspondences: List[Dict[str, Any]]):
        """معالجة دفعة من المراسلات"""
        for correspondence in correspondences:
            try:
                # Delete old embeddings first
                await qdrant_service.delete_by_correspondence_id(
                    settings.correspondence_collection,
                    correspondence["id"]
                )
                
                # Generate embeddings
                embeddings = await embedding_service.generate_embeddings_for_correspondence(correspondence)
                
                # Upsert to Qdrant
                await qdrant_service.upsert_embeddings(
                    settings.correspondence_collection,
                    embeddings
                )
                
                logger.debug(f"Synced correspondence {correspondence['id']} ({correspondence['mail_num']})")
                
            except Exception as e:
                logger.error(f"Error syncing correspondence {correspondence['id']}: {e}")
                raise
    
    async def sync_single_correspondence(self, correspondence_id: str) -> bool:
        """مزامنة مراسلة واحدة"""
        try:
            logger.info(f"Syncing single correspondence: {correspondence_id}")
            
            correspondence = await database_service.get_correspondence_by_id(correspondence_id)
            
            if not correspondence:
                logger.warn(f"Correspondence {correspondence_id} not found")
                return False
            
            # Delete old embeddings
            await qdrant_service.delete_by_correspondence_id(
                settings.correspondence_collection,
                correspondence_id
            )
            
            # Generate embeddings
            embeddings = await embedding_service.generate_embeddings_for_correspondence(correspondence)
            
            # Upsert to Qdrant
            await qdrant_service.upsert_embeddings(
                settings.correspondence_collection,
                embeddings
            )
            
            logger.info(f"Successfully synced correspondence {correspondence_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error syncing correspondence {correspondence_id}: {e}")
            return False
    
    async def delete_correspondence(self, correspondence_id: str) -> bool:
        """حذف مراسلة من Qdrant"""
        try:
            logger.info(f"Deleting correspondence {correspondence_id} from Qdrant")
            
            await qdrant_service.delete_by_correspondence_id(
                settings.correspondence_collection,
                correspondence_id
            )
            
            logger.info(f"Successfully deleted correspondence {correspondence_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error deleting correspondence {correspondence_id}: {e}")
            return False
    
    async def get_sync_statistics(self) -> Dict[str, Any]:
        """الحصول على إحصائيات المزامنة"""
        try:
            postgres_count = await database_service.get_correspondence_count()
            qdrant_count = await qdrant_service.get_collection_count(settings.correspondence_collection)
            
            return {
                "postgresql": {
                    "total": postgres_count
                },
                "qdrant": {
                    "total": qdrant_count
                },
                "synced": min(postgres_count, qdrant_count),
                "pending": max(0, postgres_count - qdrant_count)
            }
            
        except Exception as e:
            logger.error(f"Error getting sync statistics: {e}")
            raise
    
    async def rebuild_index(self) -> SyncResult:
        """إعادة بناء الفهرس من الصفر"""
        try:
            logger.info("Rebuilding index: deleting collection...")
            
            # Delete and recreate collection
            await qdrant_service.delete_collection(settings.correspondence_collection)
            qdrant_service.initialize()
            
            logger.info("Collection recreated, starting full sync...")
            
            # Perform full sync
            return await self.sync_correspondences(
                SyncRequest(type="full", batch_size=100)
            )
            
        except Exception as e:
            logger.error(f"Error rebuilding index: {e}")
            raise


# Create global sync service instance
sync_service = SyncService()


__all__ = ["sync_service", "SyncService"]

